crates/nil/src/vfs.rs

```diff
@@ -184,10 +185,14 @@ impl LineMap {
                     .map(|(_, i)| i + 1),
             )
             .collect::<Vec<_>>();
-        line_starts.push(text.len() as u32);
 
+        let ends = &line_starts[1..];
```

Comment:
This variable is only used once with another chained element. A standalone `let` does not increase readability.
My suggestion:
```rust
let start_pos_iter = line_starts.iter().copied();
let end_pos_iter = line_starts[1..].iter().copied().chain(Some(text_len));
for ((start, end), i) in start_pos_iter.zip(end_pos_iter).zip(0u32..)
```

BTW: `Some(val)` and `iter::once(val)` are interchangeable. Either is fine to me.